okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java

```diff
@@ -612,4 +613,70 @@
     assertEquals("/a/b/", base.newBuilder().addPathSegment("").addPathSegment("..").build().path());
     assertEquals("/a/b/c/", base.newBuilder().addPathSegment("").addPathSegment("").build().path());
   }
+
+  @Test public void composeQueryWithComponents() throws Exception {
+    HttpUrl base = HttpUrl.parse("http://host/");
+    HttpUrl url = base.newBuilder().addQueryParameter("a+=& b", "c+=& d").build();
+    assertEquals("http://host/?a%2B%3D%26+b=c%2B%3D%26+d", url.toString());
+    assertEquals("c+=& d", url.queryParameterValue(0));
+    assertEquals("a+=& b", url.queryParameterName(0));
+    assertEquals("c+=& d", url.queryParameter("a+=& b"));
+    assertEquals(Collections.singleton("a+=& b"), url.queryParameterNames());
+    assertEquals(Collections.singletonList("c+=& d"), url.queryParameterValues("a+=& b"));
+    assertEquals(1, url.querySize());
+    assertEquals("a+=& b=c+=& d", url.decodeQuery()); // Ambiguous! (Though working as designed.)
```

Comment:
As far as I can tell, there's applications that don't want to cut the query up into name/value pairs, and instead want a single string. This is for those applications.